windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -713,16 +726,31 @@ procedure UpdateAvailableState.HandleInstallNow;
 procedure DownloadingState.Enter;
 var
   DownloadResult: Boolean;
+  RetryCount: Integer;
 begin
   // Enter DownloadingState
   bucStateContext.SetRegistryState(usDownloading);
   { ##  for testing log that we would download }
   KL.Log('DownloadingState.HandleKmshell test code continue');
-  DownloadResult := True;
+  // DownloadResult := True;
   { End testing }
-  DownloadResult := DownloadUpdatesBackground;
-  // TODO check if keyman is running then send to Waiting Restart
-  if DownloadResult then
+  RetryCount := 0;
+  DownloadResult := False;
+
+  while (not DownloadResult) and (RetryCount < 3) do
+  begin
+    DownloadResult := DownloadUpdatesBackground;
+    if not DownloadResult then
+      Inc(RetryCount);
+  end;
+
+  if (not DownloadResult) then
+  begin
+    // Failed three times in this process return to the
+    // IdleState to wait 7 days before trying again
+    ChangeState(IdleState);
```

Comment:
Do we have to set `SRegValue_LastUpdateCheckTime` registry value to only check again in 7 days?

Also, should we log the download failure?